### PR TITLE
Giving some love to the cuttlefish escript

### PR DESCRIPTION
**Pull request by joedevivo**
- Refactored into better, more bite sized chunks
- added 'describe X' command to show docs and effective values for individual settings
- added 'effective' command to show the effective start-time riak.conf

### CODE REVIEW

File: src/cuttlefish_escript.erl [hunk 1]
```diff
@@ -51,18 +51,24 @@ print_help() ->
...
-    lists:member(help, ParsedArgs).
+parse_and_command(Args) ->
+    {ParsedArgs, Extra} = case getopt:parse(cli_options(), Args) of
+        {ok, {P, H}} -> {P, H};
+        _ -> {[help], []}
+    end,
+    Command = case {lists:member(help, ParsedArgs), Extra} of
+        {true, _} -> help;
+        {false, []} -> generate;
+        {false, [Cmd|_]} -> list_to_atom(Cmd);
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: Does it make sense to pull the command off the `Extra` list here? Then the called command won't need to match off the first element in the list.

File: src/cuttlefish_escript.erl [hunk 2]
```diff
@@ -141,6 +211,18 @@ engage_cuttlefish(ParsedArgs) ->
...
 
+    Schema = cuttlefish_schema:files(SortedSchemaFiles),
+    Schema.
+
+load_conf(ParsedArgs) ->
+    ConfFiles = proplists:get_all_values(conf_file, ParsedArgs),
+    lager:debug("ConfFiles: ~p", [ConfFiles]),
+    cuttlefish_conf:files(ConfFiles).
+
+-spec engage_cuttlefish(list()) -> {string(), string()}.
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: What type is embedded in the list?

Comment (seancribbs) [inline src/cuttlefish_escript.erl]: `[proplists:property()]` ?

File: src/cuttlefish_escript.erl [hunk 3]
```diff
@@ -75,11 +81,78 @@ main(Args) ->
...
+            print_help()
+    end.
+
+%% This shows the effective configuration, including defaults
+effective(ParsedArgs) ->
+    lager:debug("cuttlefish `meffective`", []),
+    {_, Mappings, _} = load_schema(ParsedArgs),
+    Conf = load_conf(ParsedArgs),
+
+    EffectiveConfig = cuttlefish_generator:add_defaults(Conf, Mappings),
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: This command fails at this line for me.

```
$ ./cuttlefish -e /Users/sean/dev/riak/dev/dev2/etc -d /Users/sean/dev/riak/dev/dev2/data/generated.configs -s /Users/sean/dev/riak/dev/dev2/lib -c /Users/sean/dev/riak/dev/dev2/etc/riak.conf -l debug effective
12:10:38.128 [debug] Lager installed handler error_logger_lager_h into error_logger
12:10:38.615 [debug] Lager installed handler lager_backend_throttle into lager_event
12:10:38.724 [debug] Cuttlefish set to debug level logging
12:10:38.727 [debug] cuttlefish `meffective`
12:10:38.727 [info] Application lager started on node nonode@nohost
12:10:38.727 [debug] SchemaFiles: ["/Users/sean/dev/riak/dev/dev2/lib/20-yokozuna.schema","/Users/sean/dev/riak/dev/dev2/lib/11-leveldb.schema","/Users/sean/dev/riak/dev/dev2/lib/10-multi_backend.schema","/Users/sean/dev/riak/dev/dev2/lib/06-riak_control.schema","/Users/sean/dev/riak/dev/dev2/lib/05-bitcask.schema","/Users/sean/dev/riak/dev/dev2/lib/04-riak_sysmon.schema","/Users/sean/dev/riak/dev/dev2/lib/03-riak_kv.schema","/Users/sean/dev/riak/dev/dev2/lib/02-riak_api.schema","/Users/sean/dev/riak/dev/dev2/lib/01-riak_core.schema","/Users/sean/dev/riak/dev/dev2/lib/00-riak.schema"]
12:10:38.778 [debug] ConfFiles: ["/Users/sean/dev/riak/dev/dev2/etc/riak.conf"]
escript: exception error: undefined function cuttlefish_generator:add_defaults/2
  in function  cuttlefish_escript:effective/1 (src/cuttlefish_escript.erl, line 104)
  in call from escript:run/2 (escript.erl, line 747)
  in call from escript:start/1 (escript.erl, line 277)
  in call from init:start_it/1 
  in call from init:start_em/1 
```

File: src/cuttlefish_escript.erl [hunk 4]
```diff
@@ -75,11 +81,78 @@ main(Args) ->
...
+        describe ->
+            describe(ParsedArgs, Extra);
+        Other ->
+            lager:info("Oops! All berries! ~s", [Other]),
+            print_help()
+    end.
+
+%% This shows the effective configuration, including defaults
+effective(ParsedArgs) ->
+    lager:debug("cuttlefish `meffective`", []),
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: Typo?

File: src/cuttlefish_escript.erl (hunk 3 above)
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: Works now.

File: src/cuttlefish_escript.erl [hunk 5]
```diff
@@ -75,11 +81,78 @@ main(Args) ->
...
+        catch
+            _:_ ->
+                %% I hate that I had to do this, 'cause you know...
+                %% Erlang and Strings, but actually this is ok because
+                %% sometimes there are going to be weird tuply things 
+                %% in here, so always good to fall back on ~p. 
+                %% honestly, I think this try should be built into io:format
+                ?STDOUT("~s = ~p", [Variable, Value])
+        end
+    end || {Var, Value} <- EffectiveConfig],
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl:119]: Could this be sorted before output?

File: src/cuttlefish_escript.erl [hunk 6]
```diff
@@ -75,11 +81,78 @@ main(Args) ->
...
+
+     Results = lists:filter(
+        fun(X) -> 
+            cuttlefish_util:fuzzy_variable_match(QDef, cuttlefish_mapping:variable(X))
+        end, 
+        Mappings),
+
+    case length(Results) of
+        0 -> 
+            ?STDOUT("Varible '~s' not found", [Q]);
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: Typo! This also looks strange when no query was specified.

```
Varible '' not found
```

File: src/cuttlefish_escript.erl [hunk 7]
```diff
@@ -75,11 +81,78 @@ main(Args) ->
...
+            ?STDOUT("Documentation for ~s", [string:join(cuttlefish_mapping:variable(Match), ".")]),
+            [ ?STDOUT("~s", [Line]) || Line <- cuttlefish_mapping:doc(Match)],
+            ?STDOUT("", []),
+            ?STDOUT("Datatype:      ~p", [cuttlefish_mapping:datatype(Match)]),
+            ?STDOUT("Default Value: ~p", [cuttlefish_mapping:default(Match)]),
+
+            Conf = load_conf(ParsedArgs),
+            ConfiguredValue = proplists:get_value(QDef, Conf, undefined),
+            ?STDOUT("Set Value:     ~s", [ConfiguredValue]),
+            ?STDOUT("Legacy app.config location: ~s", [cuttlefish_mapping:mapping(Match)])
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: You have beautiful formatting for the previous lines, but this one is off. Maybe shorten the text?

File: src/cuttlefish_escript.erl [hunk 8]
```diff
@@ -75,11 +81,78 @@ main(Args) ->
...
+                %% in here, so always good to fall back on ~p. 
+                %% honestly, I think this try should be built into io:format
+                ?STDOUT("~s = ~p", [Variable, Value])
+        end
+    end || {Var, Value} <- EffectiveConfig],
+    ok.
+
+%% This is the clause that dumps the docs for a single setting
+describe(ParsedArgs, ["describe"| Query]) ->
+    Q = string:join(Query, "."),
```
Comment (seancribbs) [inline src/cuttlefish_escript.erl:129]: This behavior is surprising, namely that you can provide the config names space-separated! If there is only one allowed on the command-line, I wouldn't join them. Instead maybe the pattern should be `["describe", Query|_]`?

File: src/cuttlefish_escript.erl (hunk 1 above)
Comment (joedevivo) [inline src/cuttlefish_escript.erl]: sometimes the list is empty

File: src/cuttlefish_escript.erl (hunk 2 above)
Comment (joedevivo) [inline src/cuttlefish_escript.erl]: yes. I always have a problem getting dialyzer to like proplists

File: src/cuttlefish_escript.erl (hunk 3 above)
Comment (joedevivo) [inline src/cuttlefish_escript.erl]: sorry about that. oops

File: src/cuttlefish_escript.erl (hunk 8 above)
Comment (joedevivo) [inline src/cuttlefish_escript.erl:129]: I kind of liked it this way. I did it on purpose, but I don't really remember why

File: src/cuttlefish_escript.erl (hunk 1 above)
Comment (seancribbs) [inline src/cuttlefish_escript.erl]: Yes, I know. I was suggesting `{false, [Cmd|Extra1]}` and the returning `{Command, Extra1}`

Comment (joedevivo) [inline src/cuttlefish_escript.erl]: ahh. https://twitter.com/densoneold/status/691891562

File: src/cuttlefish_escript.erl (hunk 5 above)
Comment (joedevivo) [inline src/cuttlefish_escript.erl:119]: ok

File: src/cuttlefish_escript.erl (hunk 6 above)
Comment (joedevivo) [inline src/cuttlefish_escript.erl]: ok

**Comment by joedevivo** [comment]
@seancribbs give another look now.

**Comment by seancribbs** [comment]
:+1: much improve wow

**Comment by seancribbs** [comment]
- :100: to the reversal bugfix

**Pull request merged**